engine/runtime-parser/src/main/java/org/enso/compiler/core/ir/MetadataStorage.java

```diff
@@ -214,7 +214,7 @@ public boolean equals(Object obj) {
       return true;
     }
     if (obj instanceof MetadataStorage other) {
-      return this.metadata == other.metadata;
+      return Objects.equals(this.metadata, other.metadata);
```

Comment:
This change alone is causing `StackOverflowError` in `IR.duplicate()`, @hubertp, please try to make just this change and check if you get the `StackOverflowError` as well. If so, do you have any idea how to prevent it?

Reply:
- #11015 has been proposed as a short term fix until the `StackOverflowError` & other problems are resolved.

Reply:
With 6f93f4e the `StackOverflowError` shall disappear.

Reply:
I am afraid 
- #8566

has introduced cycles in the IR graph - `TypeSignatures` metadata may probably point to `Name.Literal` that points back to the `TypeSignatures` metadata. CCing @radeusgd